src/bits.rs

```diff
@@ -0,0 +1,723 @@
+            *output_col += what_to_add.len();
+        }
+
+        let mut result = "".to_owned();
+        let (mut col, mut row) = (0, 0);
+        let mut line_ends_buffered = 0;
+        let mut output_col = 0;
+        while row < self.height() {
+            while col < self.width() {
+                let (rle_len, ch) = self.get_run(col, row, visibility);
+
+                match ch {
+                    'b' => {
+                        // Blank
+                        // TODO: if supporting diffs with this same algorithm, then need to allow
+                        // other characters to serve this purpose.
+                        if col + rle_len < self.width() {
+                            if line_ends_buffered > 0 {
+                                push(&mut result, &mut output_col, line_ends_buffered, '$');
+                                line_ends_buffered = 0;
+                            }
+                            push(&mut result, &mut output_col, rle_len, ch);
+                        }
+                    }
+                    _ => {
+                        // Non-blank
+                        if line_ends_buffered > 0 {
+                            push(&mut result, &mut output_col, line_ends_buffered, '$');
+                            line_ends_buffered = 0;
+                        }
+                        push(&mut result, &mut output_col, rle_len, ch);
+                    }
+                }
+
+                col += rle_len;
+            }
+
+            row += 1;
+            col = 0;
+            line_ends_buffered += 1;
+        }
+        push(&mut result, &mut output_col, 1, '!');
+        Pattern(result)
+    }
+
+    /// Given a starting cell at `(col, row)`, get the character at that cell, and the number of
+    /// contiguous identical cells considering only this cell and the cells to the right of it.
+    /// This is intended for exporting to RLE.
+    ///
+    /// The `visibility` parameter, if not `None`, is used to generate a run as observed by a
+    /// particular player.
+    ///
+    /// # Returns
+    ///
+    /// `(run_length, ch)`
+    ///
+    /// # Panics
+    ///
+    /// This function will panic if `col`, `row`, or `visibility` (`Some(player_id)`) are out of bounds.
+    fn get_run(&self, col: usize, row: usize, visibility: Option<usize>) -> (usize, char);
+}
+
+
+const VALID_BIT_GRID_CHARS: [char; 2] = ['b', 'o'];
+
+impl CharGrid for BitGrid {
+    /// Width in cells
+    fn width(&self) -> usize {
+        self.width_in_words() * 64
+    }
+
+    /// Height in cells
+    fn height(&self) -> usize {
+        self.0.len()
+    }
+
+    /// _visibility is ignored, since BitGrids have no concept of a player.
+    fn write_at_position(&mut self, col: usize, row: usize, ch: char, _visibility: Option<usize>) {
+        let word_col = col/64;
+        let shift = 63 - (col & (64 - 1));
+        match ch {
+            'b' => {
+                self[row][word_col] &= !(1 << shift)
```

Comment:
use `self.modify_bits_in_word(...)` for here and line 330?

Reply:
you got it! :grin: